design_notes/grouper_objects.md

```diff
@@ -0,0 +1,244 @@
+
+## Usage and impact
+
+<!-- This section describes how users of NumPy will use features described in this -->
+<!-- NEP. It should be comprised mainly of code examples that wouldn't be possible -->
+<!-- without acceptance and implementation of this NEP, as well as the impact the -->
+<!-- proposed changes would have on the ecosystem. This section should be written -->
+<!-- from the perspective of the users of NumPy, and the benefits it will provide -->
+<!-- them; and as such, it should include implementation details only if -->
+<!-- necessary to explain the functionality. -->
+
+Grouper objects
+1. Will abstract useful factorization algorithms, and
+2. Present a natural way to extend GroupBy to grouping by multiple variables: `ds.groupby(x=BinGrouper(...), t=Resampler(freq="M", ...)).mean()`.
+
+In addition, Grouper objects provide a nice interface to add often-requested grouping functionality
+1. A new `SpaceResampler` would allow specifying resampling spatial dimensions.  ([issue](https://github.com/pydata/xarray/issues/4008))
+2. `RollingTimeResampler` would allow rolling-like functionality that understands timestamps ([issue](https://github.com/pydata/xarray/issues/3216))
+3. A `QuantileBinGrouper` to abstract away `pd.cut` ([issue](https://github.com/pydata/xarray/discussions/7110))
+4. A `SeasonGrouper` and `SeasonResampler` would abstract away common annoyances with such calculations today
+   1. Support seasons that span a year-end.
+   2. Only include seasons with complete data coverage.
+   3. Allow grouping over seasons of unequal length
+   4. See [this xcdat discussion](https://github.com/xCDAT/xcdat/issues/416) for a `SeasonGrouper` like functionality:
+   5. Return results with seasons in a sensible order
+5. Weighted grouping ([issue](https://github.com/pydata/xarray/issues/3937))
+   1. Once `IntervalIndex` like objects are supported, `Resampler` groupers can account for interval lengths when resampling.
+
+## Backward Compatibility
+
+Xarray's existing grouping functionality will be exposed using two new Groupers:
+1. `UniqueGrouper` which uses `pandas.factorize`
+2. `BinGrouper` which uses `pandas.cut`
+3. `TimeResampler` which mimics pandas' `.resample`
+
+Grouping by single variables will be unaffected so that `ds.groupby('x')` will be identical to `ds.groupby(x=UniqueGrouper())`.
+Similarly, `ds.groupby_bins('x', bins=np.arange(10, 2))` will be unchanged and identical to `ds.groupby(x=BinGrouper(bins=np.arange(10, 2)))`.
+
+## Detailed description
+
+All Grouper objects will subclass from a Grouper object
+```python
+import abc
+
+class Grouper(abc.ABC):
+    @abc.abstractmethod
+    def factorize(self, group):
+        raise NotImplementedError
+
+class CustomGrouper(Grouper):
+    def factorize(self, group):
+        ...
+        return codes, group_indices, unique_coord, full_index
+
+    def weights(self, group):
+        ...
+        return weights
+```
+
+### The `factorize` method
```

Comment:
Is there a better name? `"encode"`? `"to_codes"`? `"label"`? `"to_integer_labels"`?

Reply:
`.find_groups()`? IIUC The goal of a `Grouper` is to find and extract the groups from a variable? Assuming that this method returns all the information needed to represent the groups, wouldn't it be better to encapsulate the returned items in an instance of a simple dataclass `Groups`?

And maybe rename the method argument from `group` to `var`? I find "group" a bit confusing here... It is more a variable that possibly contains groups of a certain kind.

Reply:
I didn't like `factorize` because no one knows what it means :) AFAICT pandas chose that name, and I haven't seen it used anywhere else. I think `label_groups` would be best, and most easy to understand for a new user.

Reply:
okay, that makes sense. Since this is advanced API I think we can require people to learn the new meaning, and since we're borrowing from `pandas` we're not inventing anything new (we could take `split_indices` or `to_split_indices` if we wanted to stay closer to "split-apply-combine", though).

What does other "split-apply-combine" software call this? From what I can tell, `polars` appears to call this `to_physical`. Not sure what `dplyr` calls it.